zuul-core/src/main/java/com/netflix/zuul/message/Headers.java

```diff
@@ -548,19 +633,48 @@ private void truncate(int i) {
         }
     }
 
-    private String validateField(String value) {
+    /**
+     * Checks if the given value is compliant with our RFC 7230 based check
+     */
+    private static boolean isValid(String value) {
+        if (value == null || findInvalid(value) == ABSENT) {
+            return true;
+        }
+        invalidHeaderCounter.increment();
+        return false;
+    }
+
+    /**
+     * Checks if the input value is compliant with our RFC 7230 based check
+     * Returns input value if valid, raises ZuulException otherwise
+     */
+    private static String validateField(String value) {
         if (value != null) {
-            int l = value.length();
-            for (int i = 0; i < l; i++) {
-                char c = value.charAt(i);
-                // ASCII non-control characters, per RFC 7230 but slightly more lenient
-                if (c < 31 || c == 127) {
-                    Spectator.globalRegistry().counter("zuul.header.invalid.char").increment();
-                    throw new ZuulException("Invalid header field: char " + (int) c + " in string " + value
-                            + " does not comply with RFC 7230");
-                }
+            int pos = findInvalid(value);
+            if (pos != ABSENT) {
+                invalidHeaderCounter.increment();
+                throw new ZuulException("Invalid header field: char " + (int) value.charAt(pos) + " in string " + value
+                        + " does not comply with RFC 7230");
             }
         }
         return value;
     }
+
+    /**
+     * Validated the input value based on RFC 7230 but more lenient.
+     * Currently, only ASCII control characters are considered invalid.
+     *
+     * Returns the index of first invalid character. Returns {@link #ABSENT} if absent.
+     */
+    private static int findInvalid(String value) {
+        int l = value.length();
```

Comment:
no need to extract this out, java will realize the size is a constant in the loop.

Reply:
done